src/hooks/useOnOutsideClick.tsx

```diff
@@ -0,0 +1,59 @@
+import React, {useEffect, useCallback} from 'react'
+
+export type TouchOrMouseEvent = MouseEvent | TouchEvent
+
+export type UseOnOutsideClickProps = {
+  containerRef: React.RefObject<HTMLDivElement>
+  ignoreClickRefs?: React.RefObject<HTMLElement> []
+  onClickOutside: (e: TouchOrMouseEvent) => void
+}
+
+type ShouldCallClickHandlerProps = {
+  ignoreClickRefs?: React.RefObject<HTMLElement> []
+  containerRef: React.RefObject<HTMLDivElement>
+  e: TouchOrMouseEvent
+}
+
+const shouldCallClickHandler = ({ignoreClickRefs, containerRef, e}: ShouldCallClickHandlerProps) : boolean => {
+  let shouldCallHandler = true
+
+  // don't call click handler if the mouse event was triggered by an auxiliary button (right click/wheel button/etc)
+  if (e instanceof MouseEvent && e.button > 0) {
+    shouldCallHandler = false
+  }
+
+  // don't call handler if the click happened inside of the container
+  if (containerRef && 'current' in containerRef && containerRef.current?.contains(e.target as Node)) {
+    shouldCallHandler = false
+  // don't call handler if click happened on an ignored ref
+  } else if (ignoreClickRefs) {
+    for(const ignoreRef of ignoreClickRefs) {
+      if(ignoreRef && ignoreRef.current?.contains(e.target as Node)) {
+        shouldCallHandler = false
+        // if we encounter one, break early, we don't need to go through the rest
+        break
+      }
+    }
+  }
+  return shouldCallHandler
+}
+
+
+export const useOnOutsideClick = ({containerRef, ignoreClickRefs, onClickOutside}: UseOnOutsideClickProps): void => {
+  const onOutsideClickInternal = useCallback(
+    (e: TouchOrMouseEvent) => {
+      if (shouldCallClickHandler({ignoreClickRefs, containerRef, e})){
+        onClickOutside(e)
+      }
+    }, [onClickOutside, containerRef, ignoreClickRefs]
+  )
+
+  useEffect(() => {
+    document.addEventListener('mousedown', onOutsideClickInternal)
+    document.addEventListener('touchstart', onOutsideClickInternal)
```

Comment:
Are the touch events necessary? A touch should trigger a `mousedown` event. Furthermore, is there a reason for not using the `click` event (which will also get triggered via touch)?

Reply:
Fixed. Note - it doesn't seem like there's a TS interface that is generic enough to handle either touch or mouse events so I had to keep the `TouchOrMouseEvent` stuff around

Reply:
>Fixed

Hmm, not sure what changed. I was suggesting removing the event listener for `touchstart` completely, in which case there would be no need for a unified `MouseEvent | TouchEvent` type.

Reply:
Aghh! Again, was looking at an old diff 😅

Reply:
OK, looking at the right diff this time. It's good, although I am still not convinced of a need for the unified type. It doesn't seem like there is any situation where a `TouchEvent` object is generated. A touch generates both a `MouseEvent` and a `TouchEvent`. The `MouseEvent` is passed to the `mousedown` handler (with `e.button` set to `0`), so at no point should we have to check for a `TouchEvent` instance.